File: electron/api/trades.ts
import { getData, setData } from '@storage';
import { Holding, Security } from '@types';
import { dayjsParse } from '@utils';

import { QuoteService } from '../quotes';

export interface TradeData {
  accountId: string;
  symbol: string;
  type: 'BUY' | 'SELL';
  date: string;
  quantity: number;
  price: number;
  brokerage: number;
}

/**
 * Gets the last price for the given symbol.
 *
 * @param symbol Symbol of the security
 * @returns The price as a number
 * @throws If `yahooFinance.quote()` failed or the price was not found
 */
export const lastPrice = async (symbol: string) => {
  const quoteService = new QuoteService();
  await quoteService.init();

  await quoteService.requestQuoteData([symbol], []);
  const { quote } = quoteService.getQuote(symbol);

  // Safe deconstruction (fields checked inside .getQuote())
  return quote.regularMarketPrice!;
};

/**
 * Returns the number of available units.
 *
 * @param symbol Symbol of the security
 * @param accountId Account id to check
 * @returns Number of available units
 * @throws If `symbol` does not exist
 */
export const availableUnits = async (symbol: string, accountId: string) => {
  const securities = await getData('securities');

  const security = securities.get(symbol);
  if (security === undefined) {
    throw new Error(`ERROR: Could not find data for ${symbol}.`);
  }

  return security.holdings
    .filter((entry) => entry.accountId === accountId)
    .reduce((acc, cur) => acc + Number(cur.quantity), 0);
};

/**
 * Adds a trade for a security into the application.
 *
 * @param data Data for the trade
 * @throws
 * - If `data.quantity` is zero
 * - If `data.symbol` does not exist
 * - If GST % can not be parsed from settings
 */
export const addTrade = async (data: TradeData) => {
  if (data.quantity === 0) {
    throw new Error('ERROR: Quantity is zero.');
  }

  const securities = await getData('securities');
  const settings = await getData('settings');

  const security = securities.get(data.symbol);
  if (security === undefined) {
    throw new Error(`ERROR: Could not find symbol ${data.symbol}.`);
  }

  const gstPercent = Number(settings.gstPercent);
  if (isNaN(gstPercent)) {
    throw new Error('ERROR: Could not parse GST % from settings, is NaN.');
  }

  // Perform the trade action (either buy/sell)
  const tradeAction = data.type === 'BUY' ? buyTrade : sellTrade;
  await tradeAction(data, security, gstPercent);

  await setData('securities', securities);
};

/**
 * Adds a trade for buying a security into the application.
 *
 * @param trade Data for the trade
 * @param security Object containing the security data
 * @param gstPercent GST % to use
 */
const buyTrade = async (trade: TradeData, security: Security, gstPercent: number) => {
  const { quantity, price, brokerage } = trade;
  const gst = brokerage * gstPercent / 100;
  const total = quantity * price + brokerage + gst;

  const newHolding: Holding = {
    accountId: trade.accountId,
    date: trade.date,
    quantity: trade.quantity,
    price: trade.price,
    brokerage: trade.brokerage,
    gst,
  };

  security.holdings.push(newHolding);
  security.buyHistory.push({
    tradeId: crypto.randomUUID(),
    total,
    ...newHolding,
  });

  // Ensure order is maintained
  security.buyHistory.sort((a, b) => dayjsParse(a.date).isAfter(dayjsParse(b.date)) ? 1 : -1);
};

/**
 * Adds a trade for selling a security into the application.
 *
 * @param trade Data for the trade
 * @param security Object containing the security data
 * @param gstPercent GST % to use
 */
const sellTrade = async (trade: TradeData, security: Security, gstPercent: number) => {
  // Retrieve all holdings for the account id, removing any holdings with buy dates
  // after the sell date, sorted in ascending date order (oldest first)
  const holdings = security.holdings
    .filter((holding) => holding.accountId === trade.accountId && !dayjsParse(holding.date).isAfter(dayjsParse(trade.date)))
    .sort((a, b) => (dayjsParse(a.date).isBefore(dayjsParse(b.date)) ? -1 : 1));

  // Check that the account id owns enough units for the trade
  const ownedUnits = holdings.reduce((acc, cur) => acc + Number(cur.quantity), 0);
  if (ownedUnits < trade.quantity) {
    throw new Error(`ERROR: Insufficient quantity. Required: ${trade.quantity}. Owned: ${ownedUnits}`);
  }

  const tradeId = crypto.randomUUID();
  const totalSellGst = trade.brokerage * gstPercent;

  // Keep looping until all quantity is accounted for
  let remainingQuantity = trade.quantity;
  while (remainingQuantity > 0) {
    // Retrieve next (oldest) holding
    const holding = holdings.shift()!;

    const quantitySold = Math.min(holding.quantity, remainingQuantity);
    remainingQuantity -= quantitySold;

    const buyRatio = quantitySold / holding.quantity;
    const sellRatio = quantitySold / trade.quantity;

    const appliedBuyBrokerage = buyRatio * holding.brokerage;
    const appliedSellBrokerage = sellRatio * trade.brokerage;
    const appliedBuyGst = buyRatio * holding.gst;
    const appliedSellGst = sellRatio * totalSellGst;

    const totalCost = quantitySold * holding.price + appliedBuyBrokerage + appliedBuyGst;
    const totalRevenue = quantitySold * trade.price - appliedSellBrokerage - appliedSellGst;
    const profitOrLoss = totalRevenue - totalCost;

    // Calculate the capital gain/loss
    // CGT discount (50%) applies if the owner has:
    //   1. Held onto the asset for more than 12 months, and
    //   2. Made a capital gain.
    const cgtDiscount = profitOrLoss > 0 && dayjsParse(trade.date).diff(dayjsParse(holding.date), 'year', true) > 1;
    const capitalGainOrLoss = cgtDiscount ? profitOrLoss / 2 : profitOrLoss;

    security.sellHistory.push({
      tradeId,
      accountId: trade.accountId,
      buyDate: holding.date,
      sellDate: trade.date,
      quantity: quantitySold,
      buyPrice: holding.price,
      sellPrice: trade.price,
      appliedBuyBrokerage,
      appliedSellBrokerage,
      appliedBuyGst,
      appliedSellGst,
      total: totalRevenue,
      profitOrLoss,
      capitalGainOrLoss,
      cgtDiscount,
    });

    // If the full quantity of the holding was sold, remove the holding from the security object.
    // Otherwise, remove only the amount sold from the holding.
    if (quantitySold === holding.quantity) {
      const index = security.holdings.indexOf(holding);
      if (index != -1) {
        security.holdings.splice(index, 1);
      }
    } else {
      holding.quantity -= quantitySold;
      holding.brokerage *= 1 - buyRatio;
      holding.gst *= 1 - buyRatio;
    }
  }

  // Ensure order is maintained
  security.sellHistory.sort((a, b) => dayjsParse(a.sellDate).isAfter(dayjsParse(b.sellDate)) ? 1 : -1);
};

File: src-tauri/src/commands/trades.rs
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, State};
use uuid::Uuid;

use crate::{
    store::save_securities,
    types::{BuyTrade, Holding, Security, SellTrade},
    utils::parse_date,
    AppState,
};

#[derive(Serialize, Deserialize)]
pub struct TradeData {
    pub account_id: String,
    pub symbol: String,
    pub trade_type: String, // "BUY" or "SELL"
    pub date: String,
    pub quantity: u32,
    pub price: f64,
    pub costs: f64,
}

/// Adds a trade for a security into the application.
#[tauri::command]
pub async fn add_trade(
    app: AppHandle,
    state: State<'_, AppState>,
    data: TradeData,
) -> Result<(), String> {
    if data.quantity <= 0 {
        return Err("Quantity must be > 0.".to_string()).inspect_err(|e| log::error!("{e}"));
    }

    let mut securities = state.securities.lock().await;
    let security = securities
        .get_mut(&data.symbol)
        .ok_or_else(|| format!("Could not find symbol: {}.", data.symbol))
        .inspect_err(|e| log::error!("{e}"))?;

    match data.trade_type.as_str() {
        "BUY" => buy_trade(&data, security),
        "SELL" => sell_trade(&data, security)?,
        _ => {
            return Err(format!("Unknown trade type: {}.", data.trade_type))
                .inspect_err(|e| log::error!("{e}"))
        }
    }

    save_securities(&app, &securities);

    Ok(())
}

/// Adds a buy trade for a security.
fn buy_trade(trade: &TradeData, security: &mut Security) {
    let total = trade.quantity as f64 * trade.price;

    security.holdings.push(Holding {
        account_id: trade.account_id.clone(),
        date: trade.date.clone(),
        quantity: trade.quantity,
        price: trade.price,
        costs: trade.costs,
    });

    security.buy_trades.push(BuyTrade {
        trade_id: Uuid::new_v4().to_string(),
        account_id: trade.account_id.clone(),
        date: trade.date.clone(),
        quantity: trade.quantity,
        price: trade.price,
        costs: trade.costs,
        total,
    });

    // Ensure holdings and buy trades are sorted in ascending date order (oldest first)
    security.holdings.sort_by(|a, b| a.date.cmp(&b.date));
    security.buy_trades.sort_by(|a, b| a.date.cmp(&b.date));
}

/// Adds a sell trade for a security.
fn sell_trade(trade: &TradeData, security: &mut Security) -> Result<(), String> {
    let sell_date = parse_date(&trade.date).inspect_err(|e| log::error!("{e}"))?;

    // Ensure holdings are sorted in ascending date order (oldest first)
    security.holdings.sort_by(|a, b| a.date.cmp(&b.date));

    // Check that the account owns enough units for the trade
    let owned_units: u32 = security
        .holdings
        .iter()
        .filter(|h| {
            h.account_id == trade.account_id
                && parse_date(&h.date).map_or(false, |d| d <= sell_date)
        })
        .map(|h| h.quantity)
        .sum();

    if owned_units < trade.quantity {
        return Err(format!(
            "Insufficient quantity. Required: {}. Owned: {}.",
            trade.quantity, owned_units
        ))
        .inspect_err(|e| log::error!("{e}"));
    }

    let trade_id = Uuid::new_v4().to_string();
    let mut remaining_quantity = trade.quantity;

    security.holdings.retain_mut(|holding| {
        if remaining_quantity == 0
            || holding.account_id != trade.account_id
            || parse_date(&holding.date).map_or(true, |d| d > sell_date)
        {
            return true; // keep holding (skip)
        }

        let quantity_sold = remaining_quantity.min(holding.quantity);
        remaining_quantity -= quantity_sold;

        let buy_ratio = quantity_sold as f64 / holding.quantity as f64;
        let sell_ratio = quantity_sold as f64 / trade.quantity as f64;

        let apportioned_buy_costs = buy_ratio * holding.costs;
        let apportioned_sell_costs = sell_ratio * trade.costs;

        let total_cost = quantity_sold as f64 * holding.price + apportioned_buy_costs;
        let total_revenue = quantity_sold as f64 * trade.price - apportioned_sell_costs;
        let profit_or_loss = total_revenue - total_cost;

        let buy_date = match parse_date(&holding.date) {
            Ok(date) => date,
            Err(e) => {
                log::error!("{e}");
                return true; // keep holding (skip)
            }
        };

        // CGT discount (50%) applies if the owner has:
        //  1. Held onto the asset for more than 12 months, and
        //  2. Made a capital gain.
        let held_days = (sell_date - buy_date).num_days();
        let cgt_discount = held_days > 365 && profit_or_loss > 0.0;
        let capital_gain_or_loss = if cgt_discount {
            profit_or_loss / 2.0
        } else {
            profit_or_loss
        };

        security.sell_trades.push(SellTrade {
            trade_id: trade_id.clone(),
            account_id: trade.account_id.clone(),
            buy_date: holding.date.clone(),
            sell_date: trade.date.clone(),
            quantity: quantity_sold,
            buy_price: holding.price,
            sell_price: trade.price,
            apportioned_buy_costs,
            apportioned_sell_costs,
            total: total_revenue,
            profit_or_loss,
            capital_gain_or_loss,
            cgt_discount,
        });

        if quantity_sold == holding.quantity {
            return false; // fully sold, remove holding
        } else {
            holding.quantity -= quantity_sold;
            holding.costs *= 1.0 - buy_ratio;
            return true; // partially sold, keep holding
        }
    });

    // Ensure sell trades are sorted in ascending sell date order
    security
        .sell_trades
        .sort_by(|a, b| a.sell_date.cmp(&b.sell_date));

    Ok(())
}
